src/index.test.js

```diff
@@ -336,3 +338,83 @@ describe('Methods and computeds', () => {
     })
   })
 })
+
+let goodFields = {
+  location: {
+    fields: {
+      'country.state': {
+        label: 'Dotted field name',
+        fields: {
+          zip: {},
+          name: {},
+        },
+      },
+      addresses: {
+        label: 'Array field',
+        itemField: {
+          label: 'Item field is a record',
+          fields: {
+            street: {},
+            tenants: {
+              label: 'Array field',
+              itemField: {
+                label: 'Item field is a primitive',
+              },
+            },
+          },
+        },
+      },
+    },
+  },
+}
+
+let goodValue = {
+  location: {
+    'country.state': { zip: '07016' },
+    addresses: [{ street: 'Meridian', tenants: ['John'] }],
+  },
+}
+
+describe('submit()', () => {
+  let form = null
+
+  afterEach(() => form.dispose())
+  it('fails when validation fails', async () => {
+    const submit = async () => {
+      return true
+    }
+    form = Form({ fields, value, submit })
+    await form.submit()
+    expect(form.submit.state.status).toBe('failed')
+    expect(form.submit.state.error).toBe('Validation Error')
+    expect(form.submitError).toBe('Validation Error')
+  })
+  it('succeeds when validation and submit() run', async () => {
+    const submit = () => {
+      return 'submit run'
+    }
+    form = Form({ fields: goodFields, value: goodValue, submit })
+    const result = await form.submit()
+    expect(form.submit.state.status).toBe('succeeded')
+    expect(result).toBe('submit run')
+  })
+  it('has errors when submit throws with ValidationError', async () => {
+    const submit = () => {
+      throw new ValidationError('My submit failed', {
```

Comment:
**suggestion:** Remove message

I wonder if `message` is useful here.

Reply:
I'd leave it here, as the `message` can be a summary of individual errors. Like `There are 3 rows with errors`.